src/brod_transaction.erl

```diff
@@ -0,0 +1,343 @@
+handle_call({get_producer, Topic, Partition}, _From, #state{} = OldState) ->
+  State = may_add_producer(Topic, Partition, OldState),
+  {reply, get_producer(Topic, Partition, State), State};
+
+%% @private
+handle_call(commit, _From, #state{context = CTX} = State) ->
+  ok = kpro:txn_commit(CTX),
+  {stop, normal, ok, stop_producers(State)};
+
+%% @private
+handle_call(terminate, _From, State) ->
+  {stop, normal, ok, stop_producers(State)};
+
+%% @private
+handle_call(abort, _From,
+            #state{context = CTX} = State) ->
+  ok = kpro:txn_abort(CTX),
+  {stop, normal, ok, stop_producers(State)};
+
+%% @private
+handle_call({abort, Timeout}, _From,
+            #state{context = CTX} = State) ->
+  ok = kpro:txn_abort(CTX, #{timeout => Timeout}),
+  {stop, normal, ok, stop_producers(State)};
+
+%% @private
+handle_call(stop, _From, #state{} = State) ->
+  {stop, normal, ok, stop_producers(State)};
+
+%% @private
+handle_call(Call, _From, #state{} = State) ->
+  {reply, {error, {unsupported_call, Call}}, State}.
+
+%% @private
+handle_cast(_Cast, #state{} = State) ->
+  {noreply, State}.
+
+%% @private
+handle_info(_Call, State) ->
+  {noreply, State}.
+
+%% @private
+code_change(_OldVsn, #state{} = State, _Extra) ->
+  {ok, State}.
+
+terminate(_Reason, _State) -> ok.
+
+
+%% @see start_link/2
+-spec new(pid(), config()) -> {ok, transaction()}.
+new(ClientPid, ProducerConfig) ->
+  gen_server:start_link(?MODULE,
+                        {ClientPid, make_transactional_id(), ProducerConfig},
+                        []).
+
+%% @see start_link/3
+-spec new(pid(), transactional_id(), config()) -> {ok, transaction()}.
+new(ClientPid, TxId, ProducerConfig) ->
+  gen_server:start_link(?MODULE,
+                        {ClientPid, TxId, ProducerConfig},
+                        []).
+
+%% @see start_link/3
+%% it uses a random id for the transaction
+-spec start_link(pid(), config()) -> {ok, pid()}.
+start_link(ClientPid, ProducerConfig) ->
+  gen_server:start_link(?MODULE,
+                        {ClientPid, make_transactional_id(), ProducerConfig},
+                        []).
+
+%% @doc starts a new transaction, TxId will be the id of the transaction
+%% ProducerConfig will be the configuration of the managed producers
+%% @see producer:start_link/4 for documentation about this
+-spec start_link(pid(), transactional_id(), config()) -> {ok, pid()}.
+start_link(ClientPid, TxId, ProducerConfig) ->
+  gen_server:start_link(?MODULE, {ClientPid, TxId, ProducerConfig}, []).
+
+%% @doc produces the message (key and value) to the indicated topic-partition
+%% asynchronously returning a reference to get the result.
+%% @see brod_producer:produce/3
+-spec produce(transaction(), topic(), partition(), key(), value()) ->
+  {ok, brod:call_ref()} | {error, any()}.
+produce(Transaction, Topic, Partition, Key, Value) ->
+  {ok, ProducerPid} = gen_server:call(Transaction, {get_producer, Topic, Partition}),
+  brod_producer:produce(ProducerPid, Key, Value).
```

Comment:
we can get partition leader connection from `brod_client:get_leader_connection(Client, Topic, Partition)`
then call `kpro:request_sync/3` to send the message.